### ISSUE TITLE
Support custom object key accessor in layer transitions

### ISSUE THREAD

**Issue by ilyabo** [uber/deck.gl#2570]
#### Target Use case
It's fantastic to have the support for very efficient layer transitions. Currently, there is a possibility to define [entering transitions](http://deck.gl/#/documentation/deckgl-api-reference/layers/layer?section=render-properties) for the individual objects in a layer.  However, to determine which elements are new, the array indices are used. From the docs:

> Attribute transition is performed between the values at the same index

This is a very strong limitation. In our situation changes in to data happen often and the objects must be sorted in a certain way, so we cannot put the new ones to the end of the array. Also, storing all the possibilities in the array and setting the  opacities for the "non-existing" ones to 0 would be completely impractical (we need to represent connections between pairs of locations).

#### Proposed feature
It should be possible to specify a custom key accessor function to identify objects and changes in the data passed to a layer analogously to the [D3's key parameter in selection.data()](https://github.com/d3/d3-selection#selection_data)

#### To Do List 
  - [ ] Add label and assign to milestone 
  - [ ] Coding
  - [ ] Doc update
  - [ ] What’s new update 
  - [ ] Test

**Comment by Pessimistress** [comment]
@ilyabo Thank you for the suggestion. We do plan to expand the transition feature, though I'm not sure it can be as flexible as the d3 `key` feature, mainly because of its perf implication (potentially sorting an array of 1M objects). 

Suppose we support enter/update/exit, but you cannot change the order of your objects. i.e. the transition from `A0 C0 D0` to `B1 C1 D1` will be treated as the exit of `A`, updates of `C` and `D`, and enter of `B`. However, the transition from `A0 B0 C0 D0` to `B1 C1 D1 A1` has to be treated as an exit of `A0` and an enter of `A1`. Would that be sufficient for your use case?

**Comment by ilyabo** [comment]
Thanks for your answer @Pessimistress!

We draw [arrows sorted by their weights](https://flowmap.blue/1Z6dVVFFrdooHIs8xnJ_O7eM5bhS5KscCi7G_k0jUNDI). The sorting is essential so that the larger arrows are drawn on top of thousands of smaller ones. We have interactive filters (not in the above example) which modify data and affect the sorting. So the change may well be like  `A0 B0 C0 D0` to `A1 C1 B1 D1` and it should not trigger any enter/exit transitions. That means that what you propose won't be sufficient for our case.

I understand that you have to make tradeoffs and performance is very important, but I also don't think that our case is a very special one. I think that not having the possibility to correctly identify changes will be a major blocker in many cases in which data can change.

I don't really know how it works under the hood (I believe you are using transform feedback for the animation, is that right?). But maybe you could optionally send the correct indices to the shaders along with the data to avoid re-sorting?

**Comment by Pessimistress** [comment]
In your use case, you are using order to simulate z-index. If you instead set a small `z` to the arrow positions, you should not have to reorder them (and z is transition-enabled, unlike object index).

We are indeed using transform feedback under the hood. To support reordered data, we will need to reconstruct the start buffer to match the new end buffer. I guess the perf loss is acceptable since you as the user knows what you are asking for.

**Comment by tsaiDavid** [comment]
Thanks for the awesome work on this library, wanted to say that it'd be very beneficial to have the support for identifying "new data" outside of just the array indices!

**Comment by JRVID** [comment]
@Pessimistress Just wondering if this work is still in  progress at all? I have animation requirements on a data set that constantly changes in both length and order based on incoming data from an api/ changing zoom levels and using keys to identify the objects and their new positions rather than indexes would solve my issue instantly!

**Comment by Pessimistress** [comment]
@Fabioni what options do you need other than setting `transitions` to null?

**Comment by angiehjort** [comment]
Dear @Pessimistress! i would like to +1 this issue by adding another use case
(but if you know how to solve it differently, let me know)

we are currently exploring the possibility to move from pure d3.js to Deck.gl for Gapminder charts and mark persistency is a big deal

in this example we try to animate circles across time points
https://observablehq.com/d/181f629ec299ef7a

![2024-12-09 16 40 10](https://github.com/user-attachments/assets/62a3c647-27ba-454f-b328-9a417256b513)

you can see them trading places 
my assumption is that it's because we reorder the datapoints by population (circle area), so that the smaller ones would end up in front and not be occluded by the larger ones. but then next year population changes and circles need to reorder, but then they also get reordered on x-y

in d3 we use a custom key for mark persistency, like @ilyabo suggested in the opening post

thanks for your work in open source!

**Comment by sliterok** [comment]
chatgpt created indexedList for me to avoid animating non-related objects but smh now in textLayer background is behind the actual text of it...

![Image](https://github.com/user-attachments/assets/de4cee2a-61d4-4b7b-824a-15fef4c47843)

**Comment by vgrafe** [comment]
For layers with a relatively small number of data points, I opted to create one layer per data point. I guess that there's a potential performance impact, and remember reading that deckgl can render tons of layers - this works for me for now, with around 100 data points to render and animate.

**Comment by emilgoldsmith** [comment]
+1 from my team at picogrid.com, we're building a real time tracking C2 system, and having something like this would make it much easier when getting items back from our API. We'll try out stabilizing the order of our array now realizing this limitation, and hopefully we can get that to work, but this feature would've been very helpful

**Comment by emilgoldsmith** [comment]
If we were to try contributing this feature to deck.gl is there anyone that could give us some pointers and guidance we could be in contact with?

**Comment by Fabioni** [comment]
@emilgoldsmith hey, did you make progress on this side?

**Comment by emilgoldsmith** [comment]
@Fabioni i made some decent headway at a weekend attempt on a branch, but we've ended up living with some subpar behavior up until now

**Comment by charlieforward9** [comment]
Subscribed and up to date - this would be a huge land for showing the smooth change of a trees characteristics between two data collections. A monumental animation.

<img width="384" height="304" alt="Image" src="https://github.com/user-attachments/assets/cbf7ff24-59d1-4edb-86ae-0ca3aa948e67" />

**Comment by tfk-kaare** [comment]
This is a huge +1 from us at kart.svipper.no. Vehicles enter and leave the map on most interactions, so this would really help us when refactoring vehicle animation from dom+css to canvas

**Comment by charlieforward9** [comment]
Wonder how quickly an agent would be able to bang this out in a prototype form? 

Im not utterly invested or up to speed in this domain, but I have also not seen an issue with this many thumbs up, so it appears to be one that the community - including myself - would utilize it.

Would be happy to participate in reviews

@ilyabo @Pessimistress

**Comment by chrisgervang** [comment]
Yeah, I recommend writing a short RFC and prototype for this one since there are many users we can get input from.

I'd be curious to review pros/cons of a couple high level options.

Also, given arrow + GPGPU support is now published in luma for v10, I'd be curious to learn how that changes and potentially increases the volume this feature can support
